### ISSUE TITLE
Refactor Silver's type system implementation

### ISSUE THREAD

**Issue by tedinski** [melt-umn/silver#185]
I wanted to dump this out of my head.

Silver's type system implementation could use a few improvements. This is not new features (yet) but might be necessary to do so.

First, a number of things are misnamed! We should get a renaming pass done:

```
Type -> TypeExpr
TypeList -> TypeExprs
TypeExp -> Type
```

Second, a number of type are poorly implemented. We should embraced "foreign types" for Lists, instead of the current... awfulness. (Seriously, don't look at the list extension in silver unless you want to immediate destroy it and re-do it.) This may be difficult without first solving https://github.com/melt-umn/silver/issues/25 ? Not sure.

Third, *probably* a big performance problem with the Silver compiler is that we do Unification... poorly. Right now there's a `Substitutions` list, which just screams O(N^2). We could consider alternatives. I've thought about a magically mutating `UnificationContext` type...

Finally, the major improvement is to introduce `PolyType` in addition to `Type` (which prior to the rename is currently called `TypeExp`). This does what its name suggests: represents a type that quantifies over some type variables. Then we need to actually use `PolyType` in the appropriate places. Figuring out how to do this correctly will require some good knowledge of how type systems work.

But presently, there's a few obvious places where we should be using a `PolyType`:

https://github.com/melt-umn/silver/blob/develop/grammars/silver/definition/env/DclInfo.sv#L198

```
abstract production ntDcl
top::DclInfo ::= sg::String sl::Location fn::String bound::[TyVar] ty::TypeExp ...
```

Here, `bound` are the bound variables in `ty`, so, uh, obvious change here. However, a little further up we have:

```
abstract production prodDcl
top::DclInfo ::= sg::String sl::Location ns::NamedSignature
```

This ought to be a PolyType... except that we're implicitly quantifying over all free variables of `ns`. This shouldn't be implicit. Making this non-implicit is part of what `PolyType` would be supposed to accomplish. Right now, because of this implicitness, we have to always remember to correctly call things:

```
abstract production productionReference
top::Expr ::= q::Decorated QName
{
...
  top.typerep = freshenCompletely(q.lookupValue.typerep);
}
```

The types should enforce this instead, because a `typerep` on an `Expr` should be a monotype, while the type we get when looking things up in the environment should (usually?) be a polytype.

FINALLY, I'd like to eliminate `silver:analysis:typechecking`. Type checking is too integral to how Silver works now for it to really be a separate thing. It just confuses the organization of the code to have SOME typing related stuff out there, and other stuff not.

These projects would all be good things for anyone who might want to eventually (e.g.) introduce type classes to Silver. PolyType would even be a prerequisite. (It introduces a third poly type constructor that has a constraints list...)

**Comment by tedinski** [comment]
> introduce PolyType in addition to Type

I was reminded of what might actually be the original reason I didn't do this... we have an odd case:

```
abstract production occursDcl
top::DclInfo ::= ... fnnt::String fnat::String ntty::TypeExp atty::TypeExp
```

Ugh, it's a third style entirely. Here we have:

1. The bound variables, implicit as the free variables of `ntty`
2. The nonterminal of the occurrence, in both `fnnt` and redundantly in `ntty` since only variables can be in there...
3. The attribute type's, with the constraint that the only free variables it can have at the same as `ntty`

In other words, we have one set of bound variables across two types. We shouldn't use `PolyType` here.

So my immediate thoughts here are:

1. Maybe DclInfo doesn't use PolyType in their signatures (maybe sticking to the `bound::[TyVar] ty::TypeExp` pattern), but probably should use it appropriately when returning types.
2. DclInfo could use some work to makes things consistent and clean. It seems like a bit of a mess. Why does an `ntDcl` even have a `bound` list, when all it needs is an arity? Or a TypeExpr when we always know what that is, too? Doh.

**Comment by 200sc** [comment]
#189 Finishes the first step of this refactor.

Step two points to #25 as a (possible) prerequisite, which itself points to #36 as a prerequisite, so following that chain is presumably the next step here.

**Comment by krame505** [comment]
@tedinski I'm taking a look at performing the `PolyType` refactor, in preparation for implementing type classes.  You mentioned in this issue that type classes would introduce a third constructor for `PolyType`.  To be clear, what are the other two you are proposing?  Something like
* `uniType :: (PolyType ::= Type)` for types that don't have any polymorphic type variables (needs a better name?)
* `polyType :: (PolyType ::= [TyVar] Type)` for types that do
* `constraintType :: (PolyType ::= [TyVar] [Constraint] Type)` for type class constraint types?

Anything else I should watch out for as I dive into this?

**Comment by tedinski** [comment]
I've forgotten.

Off the top of my head anymore I'm not sure why there would be more than one constructor, unless there was a significant performance difference between `monoType(t)` and `polyType([], t)`.

OutsideIn(X) (my go-to for thinking about how such a type system should work, I seem to vaguely recall they might have a more recent publication with some differences, but I've forgotten that, too) seems to only have one such constructor of poly types ("type schemes", `\sigma`, page 12, fig 1: https://www.microsoft.com/en-us/research/wp-content/uploads/2016/02/jfp-outsidein.pdf)
